stl/inc/vector

```diff
@@ -451,38 +451,44 @@ _INLINE_VAR constexpr bool
 template <class _Ty, bool = _Has_minimum_allocation_alignment<_Ty>>
 class _NODISCARD _Asan_annotater {
 private:
-    using _Alloc      = typename _Ty::allocator_type;
-    using _Pointer    = typename _Ty::pointer;
     using _Size_type  = typename _Ty::size_type;
     using _Value_type = typename _Ty::value_type;
 
 public:
-    explicit constexpr _Asan_annotater(const _Ty& _Target) noexcept : _Target_vec(_Target), _Old_last(nullptr) {}
+    _Asan_annotater(const _Asan_annotater&) = delete;
+    _Asan_annotater& operator=(const _Asan_annotater&) = delete;
+
+    explicit constexpr _Asan_annotater(const _Ty& _Target) noexcept : _Target_vec(_Target), _Asan_last(nullptr) {}
 
     _CONSTEXPR20 _Asan_annotater(const _Ty& _Target, const _Size_type _New_size)
-        : _Target_vec(_Target), _Old_last(nullptr) {
+        : _Target_vec(_Target), _Asan_last(nullptr) {
 #if _HAS_CXX20
         if (_STD is_constant_evaluated()) {
             return;
         }
 #endif // _HAS_CXX20
 
-        const auto& _My_data     = _Target_vec._Mypair._Myval2;
-        const _Pointer& _Myfirst = _My_data._Myfirst;
-        const _Pointer& _Mylast  = _My_data._Mylast;
-        const _Pointer& _Myend   = _My_data._Myend;
+        const auto& _My_data          = _Target_vec._Mypair._Myval2;
+        const _Value_type* _Raw_first = _Unfancy_maybe_null(_My_data._Myfirst);
+        const _Value_type* _Raw_last  = _Unfancy_maybe_null(_My_data._Mylast);
+        const _Value_type* _Raw_end   = _Unfancy_maybe_null(_My_data._Myend);
 
-        if (_New_size > static_cast<_Size_type>(_Myend - _Myfirst)) {
-            // New size is larger than current capacity; assume we're allocating a new block.
+        if (_New_size > static_cast<_Size_type>(_Raw_end - _Raw_first)) {
+            if (_Raw_end != _Raw_first) {
+                // New size is larger than current capacity; assume we're allocating a new block.
+                __sanitizer_annotate_contiguous_container(_Raw_first, _Raw_end, _Raw_last, _Raw_end);

```

Comment:
I do this because this block is about to be freed. If it gets free via getting passed back into the asan runtime then this work was unnecessary because asan would clean up the shadow memory annotations. However, if we're working with a custom allocator, we have no idea if the the memory is going to be recycled and we should clean up after ourselves.

Reply:
That makes perfect sense to me, and should probably be a comment since it's non-obvious.